### ISSUE TITLE
Map idea: Evolving

### ISSUE THREAD

**Issue by deferred-impact** [deferred-impact/battledudes-issue-tracker#213]
A map which, according to some rules, slowly evolves and adapts to how players behave. The simplest rule i can think of is, "this wall was destroyed" -> "increase the thickness of the walls in this area". But possibly other, more intricate rules could be devised.

Main downside is the amount of effort this would take to implement. Also, another concern is, how do we store the map state if the server is restarted - we don't want to start evolving the map from scratch, now do we.

**Comment by ghost** [comment]
How about logic like: 
Players use this route most often, lets put barriers here. 
Bullets most hit this wall, so lets just remove it. 
This is the most isolated corner, so how about we try to divert attention here by using spawn locations and objects
This area is most populated, to make it more narrow, or less spacious. 

I think the implementation would be as such:
The map would be a very basic and dull map, to begin with, **for the start of every game.** However, as the game continues, it would change, and become more interesting. 
However, the issue is, how would the map change, while being played on, without massively disrupting players' fighting. 

E.G. A player was touching a wall, while shooting another player who was shooting at them, and also missing and hitting the wall as well. Now, the game calculates that the wall is being shot at, very much, so it decides to make the walls more thicker. One problem though, the shooter who is touching the wall. If the wall decides to change, and increase thickness, at the same time while a player is touching it, wouldn't that affect their positioning? What if the physics engine send them flying away?

**Comment by deferred-impact** [comment]
> The map would be a very basic and dull map, to begin with, for the start of every game. However, as the game continues, it would change, and become more interesting.
However, the issue is, how would the map change, while being played on, without massively disrupting players' fighting.

I think this idea only really works across multiple matches, with the map evolving each time a new match starts

**Comment by ghost** [comment]
So the map would infinitely evolve, despite new matches?

**Comment by deferred-impact** [comment]
> So the map would infinitely evolve, despite new matches?

I didn't understand this question, please rephrase

**Comment by ghost** [comment]
Would the map's changes (that it makes based on its code to change) carry over new matches? In simpler terms, **would the map never reset, despite new games being played on it?**

**Comment by deferred-impact** [comment]
Reset to what?

The idea is a map that evolves every time a new match starts. What state would it even reset to. The entire point is that it keeps and accumulates the changes

**Comment by ghost** [comment]
> The entire point is that it keeps and accumulates the changes

Yes, but does it do so for every round? 
Does it ever reset to a default? I would assume the map would have a basic default layout, with some walls or whatever, it starts with, and then it builds off of that.

**Comment by deferred-impact** [comment]
> Does it ever reset to a default?

Begin round 1. Server loads some starting map layout.
Begin round 2. Server takes what happened in round 1 into account, adjusts the map layout, and uses that.
...
Begin round 1252. Server takes what happened in round 1251 into account, adjusts the map layout, and uses that.

Like, i don't undrestand the question. Are you asking if walls destroyed by players would respawn between rounds? Yes

**Comment by ghost** [comment]
Ah okay. Your description has given me my answer. The map carries over its changes infinitely. 

Dude, I don't know how to simplify it anymore. It's the most basic question ever. 
Does the map take account of what happened in the previous match, and use that for the  current match, or does it reset it do a default starting map layout every time a new round starts?

Anyways, the idea seems cool, I give a thumbs up.

**Comment by deferred-impact** [comment]
It takes into account what happened in the last match, of course. If it reset to a default layout every time, how would it be able to accumulate changes?

Your question is impossible to understand because it makes no sense. A map that always resets to the "default starting map layout" is no different from every other map we already have in the game

**Comment by ghost** [comment]
> Your question is impossible to understand because it makes no sense. A map that always resets to the "default starting map layout" is no different from every other map we already have in the game

Ok so I completely misunderstood how this algorithm works, my apologies. 
I assumed, that the map would make changes in game, as players interacted, and would make changes, **during a round.**
But I believe your idea was that the map would build off of the previous map, and change **after a round,** and the alteration would come into effect when a *new* round would start, from the previous round. 
Again, my apologies, I misunderstood the entire concept, and asked an unrelated question based off of it.

**Comment by deferred-impact** [comment]
Well the map already changes in game as players interact... that's your "everything is destructible" feature

If i was arguing for walls respawning mid game i would've said those exact words

**Comment by ghost** [comment]
By in game changes I meant like the algorithm of the map that corresponds with player interactions, and isn't destruction. Like walls becoming thicker, or respawning, or walls being removed, or things like that.
